### PR TITLE
Replace custom `AllGather` implementation

### LINKED ISSUE #10445: Remove custom `AllGatherGrad` implementation in favor of `torch.distributed`'s
### Motivation

Lightning has a utility defined for all gather with gradients here: https://github.com/PyTorchLightning/pytorch-lightning/blob/d515bcac969c2a485ada673e302bfac51f142331/pytorch_lightning/utilities/distributed.py#L200-L222

However, this is already available in torch distributed: https://github.com/pytorch/pytorch/blob/6b44e75f6bccca7acc8ec31a635f1175c265ac54/torch/distributed/nn/functional.py#L82-L94

So there's no need to redefine this in Lightning

### Pitch

Remove the custom all gather grad implementation and call torch distributed's functional API 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @rohitgr7 @akihironitta @justusschock @tchaton

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Remove custom AllGatherGrad implementation in favor of torch.distributed's

Fixes #10445
Continues #10504

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @awaelchli @carmocca @justusschock @rohitgr7 @akihironitta @tchaton

### CODE REVIEW

File: tests/deprecated_api/test_remove_1-8.py [hunk 1]
```diff
@@ -367,6 +369,16 @@ def test_v1_8_0_deprecated_lightning_optimizers():
         assert trainer.lightning_optimizers == {}
 
 
+@RunIf(skip_windows=True)
+@mock.patch("torch.distributed.all_gather")
+@mock.patch("torch.distributed.get_world_size", return_value=1)
+def test_v1_8_0_deprecated_all_gather_grad(*_):
+    tensor1 = torch.ones(1, requires_grad=True)
+
+    with pytest.deprecated_call(match="`AllGatherGrad` has been deprecated in v1.6"):
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:378]: Need to move this

File: src/lightning_lite/utilities/apply_func.py [hunk 2]
```diff
@@ -105,9 +105,4 @@ def convert_to_tensors(data: Any, device: _DEVICE) -> Any:
     # convert non-tensors
     for src_dtype, conversion_func in CONVERSION_DTYPES:
         data = apply_to_collection(data, src_dtype, conversion_func, device=device)
-
-    def _move_to_device_and_make_contiguous(t: Tensor, device: _DEVICE) -> Tensor:
-        return t.to(device).contiguous()  # type: ignore[arg-type]
```
Comment (carmocca) [inline src/lightning_lite/utilities/apply_func.py:110]: With the changes in this PR, contiguous shouldn't be necessary anymore as it was added to the distributed calls themselves.

This is technically a breaking change, but it should close the gap between pure-pytorch and pytorch-lightning.

File: src/pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -176,6 +177,40 @@ def gather_all_tensors(*args: Any, **kwargs: Any) -> Any:
     return new_gather_all_tensors(*args, **kwargs)
 
 
+class AllGatherGrad(torch.autograd.Function):
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/distributed.py:180]: In the Lite refactor, we forgot to keep this class here.

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md [hunk 4]
```diff
@@ -99,9 +99,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   * Deprecated the `pytorch_lightning.utilities.device_parser.parse_tpu_cores` in favor of `lightning_lite.accelerators.tpu.parse_tpu_cores`
   * Deprecated the `pytorch_lightning.utilities.device_parser.parse_hpus` in favor of `pytorch_lightning.accelerators.hpu.parse_hpus`
 - Deprecated duplicate `SaveConfigCallback` parameters in `LightningCLI.__init__`: `save_config_kwargs`, `save_config_overwrite` and `save_config_multifile`. New `save_config_kwargs` parameter should be used instead ([#14998](https://github.com/Lightning-AI/lightning/pull/14998))
-
-
-- Deprecated `TrainerFn.TUNING`, `RunningStage.TUNING` and `trainer.tuning` property ([#15100](https://github.com/Lightning-AI/lightning/pull/15100)
+- Deprecated `TrainerFn.TUNING`, `RunningStage.TUNING` and `trainer.tuning` property ([#15100](https://github.com/Lightning-AI/lightning/pull/15100)ç
```
Comment (awaelchli) [inline src/pytorch_lightning/CHANGELOG.md:102]:
```suggestion
- Deprecated `TrainerFn.TUNING`, `RunningStage.TUNING` and `trainer.tuning` property ([#15100](https://github.com/Lightning-AI/lightning/pull/15100))
```

File: src/pytorch_lightning/utilities/distributed.py [hunk 5]
```diff
@@ -176,6 +177,40 @@ def gather_all_tensors(*args: Any, **kwargs: Any) -> Any:
...
+    """
+
+    @staticmethod
+    def forward(  # type: ignore[override]
+        ctx: Any,
+        tensor: Tensor,
+        group: Optional["torch.distributed.ProcessGroup"] = None,
+    ) -> Tensor:
+        rank_zero_deprecation(
+            "`AllGatherGrad` has been deprecated in v1.8 and will be removed in v1.10."
```
Comment (rohitgr7) [inline src/pytorch_lightning/utilities/distributed.py:197]:
```suggestion
            "`AllGatherGrad` has been deprecated in v1.8.0 and will be removed in v1.10.0."
```

**Review by rohitgr7 — APPROVED** [review]

File: src/lightning_lite/utilities/apply_func.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning_lite/utilities/apply_func.py:110]: In which version was it introduced? Don't we need to keep it for the older PyTorch versions we support?

**Review by awaelchli — APPROVED** [review]

File: src/lightning_lite/utilities/apply_func.py (hunk 2 above)
Comment (carmocca) [reply src/lightning_lite/utilities/apply_func.py:110]: I wrote my comment above before realizing that they didn't properly fix it. It was supposed to be introduced in 1.12

So we still have this `contiguous()` call, but it's now moved to https://github.com/Lightning-AI/lightning/pull/11531/files#diff-b49e7b638737b063bc51ee0cc6727c82a3ff52bf72a40f6a49292818b2834910R176 (instead of in `convert_to_tensors`) which is the only place that needs it to the best of my knowledge

**Comment by carmocca** [comment]
The app docs failure appears because it installs `devel.txt` which includes `examples.txt` and that installs PL master: https://github.com/Lightning-AI/lightning/blob/48993cb5ed9602e2cd462adca4dcf81ff2c7bb65/requirements/app/examples

But `AllGatherGrad` is not in master, only in this PR (as expected since this PR added it)

If this PR is force-merged then the above won't be an issue because the class will then be in master. I suggest we do this and pursue a fix in a separate issue. This will be a blocker for future PRs.

**Review by akihironitta — APPROVED** [review]

File: src/lightning_lite/utilities/distributed.py [hunk 6]
```diff
@@ -164,23 +167,29 @@ def forward(  # type: ignore[override]
...
 
+def _functional_all_gather(tensor: Any, group: Any) -> Any:
+    """Compatibility layer with Windows."""
+    if sys.platform == "win32" and not _TORCH_GREATER_EQUAL_1_12:
+        # TODO: also remove `_AllGather` when support for 1.12 is dropped
+        return _AllGather.apply(tensor, group)
+    else:
+        import torch.distributed.nn
+
+        return torch.distributed.nn.functional.all_gather(tensor, group)
```
Comment (Borda) [inline src/lightning_lite/utilities/distributed.py:190]:
```suggestion

    import torch.distributed.nn

    return torch.distributed.nn.functional.all_gather(tensor, group)
```

**Review by Borda — APPROVED** [review]
